src/legacy/core_plugins/expressions/public/np_ready/public/expression_renderer.tsx

```diff
@@ -17,49 +17,95 @@
  * under the License.
  */
 
-import { useRef, useEffect } from 'react';
+import { useRef, useEffect, useState } from 'react';
 import React from 'react';
-import { ExpressionAST, IExpressionLoaderParams, IInterpreterResult } from './types';
-import { IExpressionLoader, ExpressionLoader } from './loader';
+import { EuiProgress } from '@elastic/eui';
+import { ExpressionAST, IExpressionLoaderParams } from './types';
+import { ExpressionLoader } from './loader';
 
 // Accept all options of the runner as props except for the
 // dom element which is provided by the component itself
 export interface ExpressionRendererProps extends IExpressionLoaderParams {
   className: string;
   dataAttrs?: string[];
   expression: string | ExpressionAST;
-  /**
-   * If an element is specified, but the response of the expression run can't be rendered
-   * because it isn't a valid response or the specified renderer isn't available,
-   * this callback is called with the given result.
-   */
-  onRenderFailure?: (result: IInterpreterResult) => void;
+  renderError?: (
+    errorType: 'data' | 'render',
+    error?: string | null
+  ) => React.ReactElement | React.ReactElement[];
+}
+
+interface State {
+  isLoading: boolean;
+  hasError: boolean;
```

Comment:
Would it be simpler to just change this to:

```js
interface State {
  isLoading: boolean;
  error?: Error;
}
```

Or if we really care about where the error comes from:

```js
error?: Error & { errorType: 'data' | 'render' };
```